up-l1/someip.adoc

```diff
@@ -0,0 +1,265 @@
+Overall, *SOME/IP* plays a crucial role in facilitating efficient and standardized communication between ECUs in modern automotive systems.
+
+The following specification will elaborate on how uProtocol shall be mapped to *SOME/IP* protocol so that mechatronics applications and services can be connected to non-mechatronics and cloud based applications and services.
+
+=== Open Source Libraries
+
+As of the time of writing, there are two open source projects that are available for *SOME/IP*:
+
+1. *https://github.com/COVESA/vsomeip[COVESA-vsomeip]:* C++ solution based on the original specifications developed by BMW prior to contribution to AUTOSAR
+2. *https://projects.eclipse.org/projects/automotive.sommr[Eclipse-SommR]:*  Rust implementation based on the latest AUTOSAR *SOME/IP* specifications.
+
+NOTE: Given the latest version of *SOME/IP* specifications from AUTOSAR is not open source, we cannot share or implement uProtocol using Eclipse-SommR.
+
+=== AUTOSAR References
+All references below to:
+
+- `[PRS_SOMEIP_XXXXX]` refer to link:https://some-ip.com/standards.shtml[AUTOSAR *SOME/IP* specification].
+- `[PRS_SOMEIP_SD_XXXXX]` refer to link:https://some-ip.com/standards.shtml[SOME/IP Service Discovery Protocol Specification].
+
+A search engine usually finds the correct document based on the reference number.
+
+== Transport Layer Translation
+
+=== UAttributes
+
+The following table highlights the mapping of *uAttributes* to *SOME/IP* fields that apply to all event types. For more information about *uAttributes*, please refer to link:../basics/uattributes.adoc[uAttributes Specification].
+
+
+==== Uris
+
+link:../basics/uattributes.adoc[uAttributes Specification] explains that `source` attributes defines the address of whom sent the message, while `sink` defines the destination for the message in link:../basics/uri.adoc[`UUri`] format. `UUri` is translated to *SOME/IP* attributes per <<sink-source-uri-mapping>> below.
+
+.Sink & Source URI Mapping
+[sink-source-uri-mapping, cols="1,1,2"]
+|===
+| *UUri Portion* | *SOME/IP Field* | *Description*
+a| `authority_name` | IP address & port
+a| IP address & port (destination endpoint) of the mDevice.
+
+NOTE: `vsomeip` hides IP endpoint information and uses `Service ID`/`Instance ID` instead.
+
+a| `ue_id` a| `Instance ID` \| `Service ID`
+a| `Service ID` *MUST* be encoded in the lower 16 bits of `resource_id`.
+
+`Instance ID` *MAY* be encoded in the upper 16 bits of the 32 bit `ue_id`.
+If the upper 16 bits are `0x0000` (a reserved instance value), `Instance ID` is assumed to be `1`.
+
+NOTE: Explicitly setting `Instance ID` to `0x0001` *SHOULD* be avoided as it increases the message size.
+
+a| `ue_version_major` a| `Interface Version`
```

Comment:
As per my understanding,
uProtocol "ue_version_major" is mapped to Service Major Version in someip. What about minor version?
In someip world, minor version is also required at the time of offer service at server side and request service at client side. We will have a problem when mE would try to offer a service with different minor version and Streamer will need to request the same service with that specific minor version. 
I think, we need a mapping of ue_minor_version to map it to vsomeip Service Minor version as well.
Please confirm

Reply:
`minor version` is needed indeed , if omitted vsomeip assumes `vsomeip::DEFAULT_VERSION=0x00000000`.
The same applies for `major version` (assumed `0x00`, if not provided)
As minor is not available in uProtocol, I suppose it should be a transport configuration (especially for offered services).
For requested services we can get away with `vsomeip::ANY_MINOR`

The transport *MUST* know in advance which services/instances/major/minor and events it *MAY* interact with and pre-register them in the vsomeip router. It could be possible to request the event in e.g. `on_available` handler, but that may block the dispatcher thread for too long and you'll get vsomeip warnings.